### ISSUE TITLE
[Core Feature] Support for enum types used in dataclasses

### ISSUE THREAD

**Issue by palchicz** [flyteorg/flyte#1514]
**Motivation: Why do you think this is important?**
Since [dataclasses](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/type_system/custom_objects.html) are a supported type and [Enums](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/type_system/enums.html) are a supported type, one would expect that dataclasses that reference a enum should also be supported. Instead, the protobuf fails to form properly and I get the following warning from the command

`pyflyte --pkgs <PACKAGE_NAME> serialize --image <IMAGE_NAME>  workflows`

> WARNING:flytekit:failed to extract schema for object <class 'xxx.yyy.zzz'>, (will run schemaless) error: Currently do not support JSON schema for enums loaded by value

**Goal: What should the final outcome look like, ideally?**
I should be able to define a dataclass and enum like the following:

```python
class Animal(Enum):
    DOG = "dog"
    CAT = "cat"

@dataclass_json
@dataclass
class Person(object):
    pet: Animal = Animal.DOG
```

**Describe alternatives you've considered**
I've considered implementing a [new custom type](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/extend_flyte/run_custom_types.html#advanced-custom-types). 

**[Optional] Propose: Link/Inline OR Additional context**
This might have its root cause in this dependency https://github.com/fuhrysteve/marshmallow-jsonschema/blob/master/marshmallow_jsonschema/base.py#L228

**Comment by kumare3** [comment]
@palchicz this is because of the underlying dataclass-json library. We are actually working on supporting #1362 Pickle types. But, pickle is not native to other languages and wont show up in the UI.
Whats the usecase?

**Comment by palchicz** [comment]
Thanks for the quick response! We have a collection of ~10 workflow inputs that are shared among several workflows. These inputs are a collection of closely related parameters, and they always show up together across different workflows, so it makes sense to group them together. Consolidating them into a single dataclass made sense, however, one of these shared inputs is an enum and thus we run into the problem stated above.

For the time being, we can still use the dataclass to group 9/10 of the parameters together so we don't have to repeatedly and explicitly lay them out for each workflow. It's just awkward to have the one odd parameter that we *do* need to add to the workflow definition explicitly just because it's an enum.  


Here are some code example to help clarify what I said above:

```python
class EnumParam(Enum):
    OPTION_1 = "option 1"
    OPTION_2 = "option 2"

# Naive approach
@workflow
def workflow_1(param_1: int, param_2: int, ...., param_n: EnumParam) -> int:
   ...

@workflow
def workflow_2(param_1: int, param_2: int, ...., param_n: EnumParam) -> str:
   ...

# Envisioned Approach
@dataclass_json
@dataclass
class WorkflowParameters(object):
    param_1: int = 0
    param_2: int = 0
    ...
    param_n: EnumParam = EnumParam.OPTION_1

@workflow
def workflow_1(workflow_params: WorkflowParameters) -> int:
   ...

@workflow
def workflow_2(workflow_params: WorkflowParameters) -> str:
   ...

# Workaround 
@dataclass_json
@dataclass
class WorkflowParameters(object):
    param_1: int = 0
    param_2: int = 0
    ...


@workflow
def workflow_1(workflow_params: WorkflowParameters, param_n: EnumParam) -> int:
   ...

@workflow
def workflow_2(workflow_params: WorkflowParameters, param_n: EnumParam) -> str:
   ...
```

**Comment by kumare3** [comment]
@palchicz appreciate giving more clarification. your request absolutely makes sense. One follow up question, do you use the Enum Form in the UI - as shown below?

![Screen Shot 2021-09-23 at 4 32 48 PM](https://user-images.githubusercontent.com/16888709/134597050-6571f2de-3e61-4cd8-8cc7-495ca627ddf8.png)

This UI feature will not work with dataclasses with nested enums, that is because dataclass_json does not support enum schema extraction AFAICT. But JSON actually supports that. Let us try that.

In anycase enums should be supported by dataclasses, and I think we just need to enable one more library for that? 
marshmallow_enum - https://github.com/fuhrysteve/marshmallow-jsonschema/blob/master/marshmallow_jsonschema/base.py#L22-L27.

Can you try installing the library and help us debug, and please jump into our slack channel - [here](https://docs.google.com/forms/d/e/1FAIpQLScWPJZncyL-bBCWbuQ9HWuMzA8-r2RJc__CSWJoAYUQADN-BQ/viewform) and we can work on it with you.

**Comment by palchicz** [comment]
> One follow up question, do you use the Enum Form in the UI - as shown below?
That would be ideal, but not a deal breaker. At the very least there isn't any urgency, so maybe the dataclass_json people will add that feature before too long :smiley: :crossed_fingers: 

I can certainly jump in and help. I should have some time next week. I'll join the slack channel like you suggested.

**Comment by kumare3** [comment]
> > One follow up question, do you use the Enum Form in the UI - as shown below?
> 
> 
> 
> That would be ideal, but not a deal breaker. At the very least there isn't any urgency, so maybe the dataclass_json people will add that feature before too long :smiley: :crossed_fingers: 
> 
> 
> 
> I can certainly jump in and help. I should have some time next week. I'll join the slack channel like you suggested. 

Awesome, see you soon. Also, if the form is not required- enums should be supported. I think the enum marshmallow library might work

**Comment by palchicz** [comment]
> In anycase enums should be supported by dataclasses, and I think we just need to enable one more library for that?
marshmallow_enum - https://github.com/fuhrysteve/marshmallow-jsonschema/blob/master/marshmallow_jsonschema/base.py#L22-L27.

> Can you try installing the library and help us debug

I took a stab at debugging this problem and noticed that `marshmallow_enum` is already installed. The issue is this line [here](https://github.com/fuhrysteve/marshmallow-jsonschema/blob/81eada1a0c42ff67de216923968af0a6b54e5dcb/marshmallow_jsonschema/base.py#L227) in the `mashmallow_jsonschema` library which raises an exception when trying to convert the marshmallow enum field into a json schema enum. If you trace the code, it wouldn't get to this point unless `marshmallow_enum` was installed.

The exact reason it raises an exception is because the marshmallow Enum field has the property `field.load_by == LoadDumpOptions.value` (which gets set [here](https://github.com/lidatong/dataclasses-json/blob/master/dataclasses_json/mm.py#L263)). The matshmallow_json libraby is probably a little too restrictive in that it just errors out whenever it encounters this case. It should probably allow strings even when the field was loaded by value.

Anyways, that's as far as I got. No solution to propose, but that at least seems to be the issue.
